cloudinit/distros/__init__.py

```diff
@@ -791,6 +798,89 @@ def chpasswd(self, plist_in: list, hashed: bool):
         cmd = ["chpasswd"] + (["-e"] if hashed else [])
         subp.subp(cmd, payload)
 
+    def is_doas_rule_valid(self, user, rule):
+        rule_pattern = (
+            r"^(?:permit|deny)"
+            r"(?:\s+(?:nolog|nopass|persist|keepenv|setenv \{[^}]+\})+)*"
+            r"\s+([a-zA-Z0-9_]+)+"
+            r"(?:\s+as\s+[a-zA-Z0-9_]+)*"
+            r"(?:\s+cmd\s+[^\s]+(?:\s+args\s+[^\s]+(?:\s*[^\s]+)*)*)*"
+            r"\s*$"
+        )
+
+        LOG.debug(
+            "Checking if user '%s' is referenced in doas rule %r", user, rule
+        )
+
+        valid_match = re.search(rule_pattern, rule)
+        if valid_match:
+            LOG.debug(
+                "User '%s' referenced in doas rule", valid_match.group(1)
+            )
+            if valid_match.group(1) == user:
+                LOG.debug("Correct user is referenced in doas rule")
+                return True
+            else:
+                LOG.debug(
+                    "Incorrect user '%s' is referenced in doas rule",
+                    valid_match.group(1),
+                )
+                return False
+        else:
+            LOG.debug("doas rule does not appear to reference any user")
+            return False
+
+    def write_doas_rules(self, user, rules, doas_file=None):
+        if not doas_file:
+            doas_file = self.doas_fn
+
+        if isinstance(rules, (list, tuple)):
+            for rule in rules:
+                if not self.is_doas_rule_valid(user, rule):
+                    msg = (
+                        "Invalid doas rule %r for user '%s',"
+                        " not writing any doas rules for user!" % (rule, user)
+                    )
+                    LOG.error(msg)
+                    return
+        elif isinstance(rules, str):
```

Comment:
This should never be true with the updated jsonschema, right?

Reply:
Right I missed that, I fixed this and a similar section just after it.